tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/ScalaJSClassEmitter.scala

```diff
@@ -745,14 +858,27 @@ private[scalajs] final class ScalaJSClassEmitter(
     envFieldDef("d", className, typeData)
   }
 
+  def genSetTypeTag(tree: LinkedClass): js.Tree = {
+    import TreeDSL._
+
+    implicit val pos = tree.pos
+
+    assert(tree.kind.isClass)
+    val tag = typeTags(tree)
+
+    encodeClassVar(tree.name.name).prototype DOT "$typeTag" :=
+        js.IntLiteral(tag)
+  }
+
   def genSetTypeData(tree: LinkedClass): js.Tree = {
     import TreeDSL._
 
     implicit val pos = tree.pos
+    val tag = typeTags(tree)
 
     assert(tree.kind.isClass)
 
-    encodeClassVar(tree.name.name).prototype DOT "$classData" :=
+    js.BracketSelect(envField("ClassData"), js.IntLiteral(tag)) :=
       envField("d", tree.name.name)
```

Comment:
It's probably best to keep it the way you did in this PR. But I'm noting here to remember that we should try to assign the entire `ClassData` array in one go as an array literal, rather than building it up like this. This would be smaller.